src/RequestParser.cpp

```diff
@@ -0,0 +1,551 @@
+ *
+ * @param headerName The HTTP header field name to be validated.
+ * @throws std::runtime_error If the header name contains invalid characters or
+ *         ends with whitespace, an error is thrown and the error code is set to 400.
+ */
+void	RequestParser::checkHeaderName(const std::string& headerName)
+{
+	if (isspace(headerName[headerName.size() - 1])){
+		m_errorCode = 400;
+		throw std::runtime_error(ERR_HEADER_COLON_WHITESPACE);
+	}
+	for (size_t i = 0; i < headerName.size(); i++) {
+		if (!isValidHeaderFieldNameChar(headerName[i])) {
+			m_errorCode = 400;
+			throw std::runtime_error(ERR_HEADER_NAME_INVALID_CHAR);
+	 	}
+	}
+}
+
+/**
+ * @brief Checks the validity of the Content-Length header field.
+ *
+ * This function verifies the validity of the Content-Length header field in an HTTP request.
+ * It ensures that:
+ * - There is only one Content-Length header or multiple with the same value.
+ * - The value of Content-Length is a valid numeric string.
+ * If the header is invalid or there are multiple Content-Length headers with conflicting values,
+ * an appropriate error code is set, and a `std::runtime_error` is thrown.
+ *
+ * @param headerName The name of the HTTP header field.
+ * @param headerValue The value associated with the HTTP header field.
+ * @throws std::runtime_error If there are multiple conflicting Content-Length
+ *         headers or if the Content-Length value is invalid, an error is thrown
+ *         and the error code is set to 400.
+ */
+void	RequestParser::checkContentLength(const std::string& headerName, std::string& headerValue)
+{
+	if (headerName == "Content-Length") {
+		if (headerValue.empty()) {
+			m_errorCode = 400;
+			throw std::runtime_error(ERR_INVALID_CONTENT_LENGTH);
+		}
+		if (m_request.headers.find("Content-Length") != m_request.headers.end()
+			&& m_request.headers["Content-Length"] != headerValue) {
+				m_errorCode = 400;
+				throw std::runtime_error(ERR_MULTIPLE_CONTENT_LENGTH_VALUES);
+		}
+
+		std::vector<std::string> strValues = split(headerValue, ',');
+		std::vector<double>	numValues;
+		for (size_t i = 0; i < strValues.size(); i++) {
+			char 	*endptr;
+			double	contentLength = strtod(strValues[i].c_str(), &endptr);
+			if (!contentLength || *endptr != '\0') {
+				m_errorCode = 400;
+				throw std::runtime_error(ERR_INVALID_CONTENT_LENGTH);
+			}
+			numValues.push_back(contentLength);
+			if (i != 0 && contentLength != numValues[i - 1]) {
+				m_errorCode = 400;
+				throw std::runtime_error(ERR_MULTIPLE_CONTENT_LENGTH_VALUES);
+			}
+		}
+		m_request.hasBody = true;
+		headerValue = strValues[0];
+	}
+}
+
+/**
+ * @brief Checks the validity of Transfer-Encoding header field.
+ *
+ * This function verifies the validity of the Transfer-Encoding header field in an HTTP request.
+ * It ensures that:
+ * - If the Transfer-Encoding header exists, it is not empty.
+ * - The last encoding in the Transfer-Encoding list is "chunked".
```

Comment:
If a Transfer-Encoding header field
is present in a request and the chunked transfer coding is not
the final encoding, the message body length cannot be determined
reliably; the server MUST respond with the 400 (Bad Request)
status code and then close the connection.

Reply:
Straight from the source - nice!

This seems to be a special case for 400, because we also have to close the connection. 
Quick googlin found this [Stackoverflow Answer](https://stackoverflow.com/a/25133723): "When the server receives a 400 Bad Request, it is going to send the response with the keep-alive header because if the client feels like sending another request, then they can use a pre-existing connection" 
But in this case we have to close?

Reply:
Yes, it seems so. It is indeed interesting that the behaviour varies for the 400 error. We may need another flag then to indicate whether the connection should be closed after an error. I will then scan again all error cases to determine whether the connection shall be preserved or closed.

Reply:
Ok should we open an issue to track this?

Reply:
Not necessary, I will just implement it in the changes of this PR.

Reply:
Added bool to HTTPRequest to check whether a connection needs to be closed.